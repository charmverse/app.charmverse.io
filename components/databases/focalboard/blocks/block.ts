// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import difference from 'lodash/difference'

import {Utils} from '../utils'

const contentBlockTypes = ['text', 'image', 'divider', 'checkbox'] as const
const blockTypes = [...contentBlockTypes, 'board', 'view', 'card', 'comment', 'unknown'] as const
type ContentBlockTypes = typeof contentBlockTypes[number]
type BlockTypes = typeof blockTypes[number]

interface BlockPatch {
    workspaceId?: string
    parentId?: string
    rootId?: string
    schema?: number
    type?: BlockTypes
    title?: string
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    updatedFields?: Record<string, any>
    deletedFields?: string[]
    deleteAt?: number
}

interface Block {
    id: string
    workspaceId: string
    parentId: string
    rootId: string
    createdBy: string
    modifiedBy: string

    schema: number
    type: BlockTypes
    title: string
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    fields: Record<string, any>

    createAt: number
    updateAt: number
    deleteAt: number
}

function createBlock(block?: Block): Block {
    const now = Date.now()
    return {
        id: block?.id || Utils.createGuid(Utils.blockTypeToIDType(block?.type)),
        schema: 1,
        workspaceId: block?.workspaceId || '',
        parentId: block?.parentId || '',
        rootId: block?.rootId || '',
        createdBy: block?.createdBy || '',
        modifiedBy: block?.modifiedBy || '',
        type: block?.type || 'unknown',
        fields: block?.fields ? {...block?.fields} : {},
        title: block?.title || '',
        createAt: block?.createAt || now,
        updateAt: block?.updateAt || now,
        deleteAt: block?.deleteAt || 0,
    }
}

// createPatchesFromBlock creates two BlockPatch instances, one that
// contains the delta to update the block and another one for the undo
// action, in case it happens
function createPatchesFromBlocks(newBlock: Block, oldBlock: Block): BlockPatch[] {
    const newDeletedFields = difference(Object.keys(newBlock.fields), Object.keys(oldBlock.fields))
    const newUpdatedFields: Record<string, any> = {}
    const newUpdatedData: Record<string, any> = {}
    Object.keys(newBlock.fields).forEach((val) => {
        if (oldBlock.fields[val] !== newBlock.fields[val]) {
            newUpdatedFields[val] = newBlock.fields[val]
        }
    })
    Object.keys(newBlock).forEach((val) => {
        if (val !== 'fields' && (oldBlock as any)[val] !== (newBlock as any)[val]) {
            newUpdatedData[val] = (newBlock as any)[val]
        }
    })

    const oldDeletedFields = difference(Object.keys(oldBlock.fields), Object.keys(newBlock.fields))
    const oldUpdatedFields: Record<string, any> = {}
    const oldUpdatedData: Record<string, any> = {}
    Object.keys(oldBlock.fields).forEach((val) => {
        if (oldBlock.fields[val] !== newBlock.fields[val]) {
            oldUpdatedFields[val] = oldBlock.fields[val]
        }
    })
    Object.keys(oldBlock).forEach((val) => {
        if (val !== 'fields' && (oldBlock as any)[val] !== (newBlock as any)[val]) {
            oldUpdatedData[val] = (oldBlock as any)[val]
        }
    })

    return [
        {
            ...newUpdatedData,
            updatedFields: newUpdatedFields,
            deletedFields: oldDeletedFields,
        },
        {
            ...oldUpdatedData,
            updatedFields: oldUpdatedFields,
            deletedFields: newDeletedFields,
        },
    ]
}

export type {ContentBlockTypes, BlockTypes}
export {blockTypes, contentBlockTypes, Block, BlockPatch, createBlock, createPatchesFromBlocks}
