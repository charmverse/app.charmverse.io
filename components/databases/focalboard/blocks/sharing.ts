// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

interface ISharing {
    id: string,
    enabled: boolean,
    token: string,
    modifiedBy?: string,
    updateAt?: number,
}

export {ISharing}
