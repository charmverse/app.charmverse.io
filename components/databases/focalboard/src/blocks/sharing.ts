// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export interface ISharing {
    id: string,
    enabled: boolean,
    token: string,
    updatedBy?: string,
    updatedAt?: number,
}
