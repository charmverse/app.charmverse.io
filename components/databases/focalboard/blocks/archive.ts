// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Block} from './block'

interface ArchiveHeader {
    version: number
    date: number
}

interface ArchiveLine {
    type: string,
    data: unknown,
}

// This schema allows the expansion of additional line types in the future
interface BlockArchiveLine extends ArchiveLine {
    type: 'block',
    data: Block
}

class ArchiveUtils {
    static buildBlockArchive(blocks: readonly Block[]): string {
        const header: ArchiveHeader = {
            version: 1,
            date: Date.now(),
        }

        const headerString = JSON.stringify(header)
        let content = headerString + '\n'
        for (const block of blocks) {
            const line: BlockArchiveLine = {
                type: 'block',
                data: block,
            }
            const lineString = JSON.stringify(line)
            content += lineString
            content += '\n'
        }

        return content
    }

    static parseBlockArchive(contents: string): Block[] {
        const blocks: Block[] = []
        const allLineStrings = contents.split('\n')
        if (allLineStrings.length >= 2) {
            const headerString = allLineStrings[0]
            const header = JSON.parse(headerString) as ArchiveHeader
            if (header.date && header.version >= 1) {
                const lineStrings = allLineStrings.slice(1)
                let lineNum = 2
                for (const lineString of lineStrings) {
                    if (!lineString) {
                        // Ignore empty lines, e.g. last line
                        continue
                    }
                    const line = JSON.parse(lineString) as ArchiveLine
                    if (!line || !line.type || !line.data) {
                        throw new Error(`ERROR parsing line ${lineNum}`)
                    }
                    switch (line.type) {
                    case 'block': {
                        const blockLine = line as BlockArchiveLine
                        const block = blockLine.data
                        blocks.push(block)
                        break
                    }
                    }

                    lineNum += 1
                }
            } else {
                throw new Error('ERROR parsing header')
            }
        }

        return blocks
    }
}

export {ArchiveHeader, ArchiveLine, BlockArchiveLine, ArchiveUtils}
