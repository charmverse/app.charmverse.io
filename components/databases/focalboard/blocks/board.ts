// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils, IDType} from '../utils'

import {Block, createBlock} from './block'
import {Card} from './card'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IPropertyOption[]
}

type BoardFields = {
    icon: string
    description: string
    showDescription?: boolean
    isTemplate?: boolean
    cardProperties: IPropertyTemplate[]
    columnCalculations: Record<string, string>
}

type Board = Block & {
    fields: BoardFields
}

function createBoard(block?: Block): Board {
    let cardProperties: IPropertyTemplate[] = []
    const selectProperties = cardProperties.find((o) => o.type === 'select')
    if (!selectProperties) {
        const property: IPropertyTemplate = {
            id: Utils.createGuid(IDType.BlockID),
            name: 'Status',
            type: 'select',
            options: [],
        }
        cardProperties.push(property)
    }

    if (block?.fields.cardProperties) {
        // Deep clone of card properties and their options
        cardProperties = block?.fields.cardProperties.map((o: IPropertyTemplate) => {
            return {
                id: o.id,
                name: o.name,
                type: o.type,
                options: o.options ? o.options.map((option) => ({...option})) : [],
            }
        })
    }

    return {
        ...createBlock(block),
        type: 'board',
        fields: {
            showDescription: block?.fields.showDescription || false,
            description: block?.fields.description || '',
            icon: block?.fields.icon || '',
            isTemplate: block?.fields.isTemplate || false,
            columnCalculations: block?.fields.columnCalculations || [],
            cardProperties,
        },
    }
}

type BoardGroup = {
    option: IPropertyOption
    cards: Card[]
}

export {Board, PropertyType, IPropertyOption, IPropertyTemplate, BoardGroup, createBoard}
