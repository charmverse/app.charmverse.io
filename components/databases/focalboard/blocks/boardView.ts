// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Block, createBlock} from './block'
import {FilterGroup, createFilterGroup} from './filterGroup'

type IViewType = 'board' | 'table' | 'gallery' | 'calendar'
type ISortOption = { propertyId: '__title' | string, reversed: boolean }

type KanbanCalculationFields = {
    calculation: string
    propertyId: string
}

type BoardViewFields = {
    viewType: IViewType
    groupById?: string
    dateDisplayPropertyId?: string
    sortOptions: ISortOption[]
    visiblePropertyIds: string[]
    visibleOptionIds: string[]
    hiddenOptionIds: string[]
    collapsedOptionIds: string[]
    filter: FilterGroup
    cardOrder: string[]
    columnWidths: Record<string, number>
    columnCalculations: Record<string, string>
    kanbanCalculations: Record<string, KanbanCalculationFields>
    defaultTemplateId: string
}

type BoardView = Block & {
    fields: BoardViewFields
}

function createBoardView(block?: Block): BoardView {
    return {
        ...createBlock(block),
        type: 'view',
        fields: {
            viewType: block?.fields.viewType || 'board',
            groupById: block?.fields.groupById,
            dateDisplayPropertyId: block?.fields.dateDisplayPropertyId,
            sortOptions: block?.fields.sortOptions?.map((o: ISortOption) => ({...o})) || [],
            visiblePropertyIds: block?.fields.visiblePropertyIds?.slice() || [],
            visibleOptionIds: block?.fields.visibleOptionIds?.slice() || [],
            hiddenOptionIds: block?.fields.hiddenOptionIds?.slice() || [],
            collapsedOptionIds: block?.fields.collapsedOptionIds?.slice() || [],
            filter: createFilterGroup(block?.fields.filter),
            cardOrder: block?.fields.cardOrder?.slice() || [],
            columnWidths: {...(block?.fields.columnWidths || {})},
            columnCalculations: {...(block?.fields.columnCalculations) || {}},
            kanbanCalculations: {...(block?.fields.kanbanCalculations) || {}},
            defaultTemplateId: block?.fields.defaultTemplateId || '',
        },
    }
}

function sortBoardViewsAlphabetically(views: BoardView[]): BoardView[] {
    // Strip leading emoji to prevent unintuitive results
    return views.map((v) => {
        return {view: v, title: v.title.replace(/^\p{Emoji}*\s*/u, '')}
    }).sort((v1, v2) => v1.title.localeCompare(v2.title)).map((v) => v.view)
}

export {BoardView, IViewType, ISortOption, sortBoardViewsAlphabetically, createBoardView, KanbanCalculationFields}
