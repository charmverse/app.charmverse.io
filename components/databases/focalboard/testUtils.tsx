// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'
import {DndProvider} from 'react-dnd'
import {HTML5Backend} from 'react-dnd-html5-backend'
import configureStore, {MockStoreEnhanced} from 'redux-mock-store'
import {Middleware} from 'redux'

import {Block} from './blocks/block'

export const wrapIntl = (children?: React.ReactNode): JSX.Element => <IntlProvider locale='en'>{children}</IntlProvider>
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
    return (
        <DndProvider backend={HTML5Backend}>
            {wrapIntl(children)}
        </DndProvider>
    )
}

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()
        range.getBoundingClientRect = jest.fn()
        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }
        return range
    }
}
export function mockMatchMedia(result: {matches: boolean}): void {
    // We check if system preference is dark or light theme.
    // This is required to provide it's definition since
    // window.matchMedia doesn't exist in Jest.
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation(() => {
            return result

            // return ({
            //     matches: true,
            // })
        }),
    })
}

export function mockStateStore(middleware:Middleware[], state:unknown): MockStoreEnhanced<unknown, unknown> {
    const mockStore = configureStore(middleware)
    return mockStore(state)
}

export type BlocksById<BlockType> = {[key: string]: BlockType}

export function blocksById<BlockType extends Block>(blocks: Array<BlockType>): BlocksById<BlockType> {
    return blocks.reduce((res, block) => {
        res[block.id] = block
        return res
    }, {} as BlocksById<BlockType>)
}
