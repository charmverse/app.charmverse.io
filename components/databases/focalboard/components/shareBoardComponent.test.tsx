// @ts-nocheck
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {act, render, screen} from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import React from 'react'
import {MemoryRouter} from 'react-router'
import {mocked} from 'ts-jest/utils'

import {ISharing} from '../blocks/sharing'
import {TestBlockFactory} from '../test/testBlockFactory'
import {wrapDNDIntl} from '../testUtils'
import client from '../octoClient'
import {Utils} from '../utils'

import ShareBoardComponent from './shareBoardComponent'

jest.useFakeTimers()

const boardId = '1'
const workspaceId: string|undefined = boardId
const viewId = boardId

jest.mock('../octoClient')
jest.mock('../utils')

const mockedOctoClient = mocked(client, true)
const mockedUtils = mocked(Utils, true)

let params = {}
jest.mock('react-router', () => {
    const originalModule = jest.requireActual('react-router')

    return {
        ...originalModule,
        useRouteMatch: jest.fn(() => {
            return {
                url: 'http://localhost/',
                path: '/',
                params,
                isExact: true,
            }
        }),
    }
})
const board = TestBlockFactory.createBoard()
board.id = boardId

describe('src/components/shareBoardComponent', () => {
    const w = (window as any)
    const oldBaseURL = w.baseURL

    beforeEach(() => {
        jest.clearAllMocks()
        mockedUtils.buildURL.mockImplementation((path) => (w.baseURL || '') + path)

        params = {
            boardId,
            viewId,
            workspaceId,
        }
    })

    afterEach(() => {
        w.baseURL = oldBaseURL
    })

    test('should match snapshot', async () => {
        mockedOctoClient.getSharing.mockResolvedValue(undefined)
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
    test('should match snapshot with sharing', async () => {
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
    test('should match snapshot with sharing and without workspaceId', async () => {
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        params = {
            boardId,
            viewId,
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
    test('return shareBoardComponent and click Copy link', async () => {
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container: Element | undefined
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        const copyLinkElement = screen.getByRole('button', {name: 'Copy link'})
        expect(copyLinkElement).toBeDefined()
        userEvent.click(copyLinkElement)
        expect(mockedUtils.copyTextToClipboard).toBeCalledTimes(1)
        expect(container).toMatchSnapshot()
    })
    test('return shareBoardComponent and click Regenerate token', async () => {
        window.confirm = jest.fn(() => {
            return true
        })
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container: Element | undefined
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        sharing.token = 'anotherToken'
        mockedUtils.createGuid.mockReturnValue('anotherToken')
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        mockedOctoClient.setSharing.mockResolvedValue(true)
        const regenerateTokenElement = screen.getByRole('button', {name: 'Regenerate token'})
        expect(regenerateTokenElement).toBeDefined()
        userEvent.click(regenerateTokenElement)
        await act(async () => {
            jest.runOnlyPendingTimers()
        })
        expect(mockedOctoClient.setSharing).toBeCalledTimes(1)
        expect(container).toMatchSnapshot()
    })
    test('return shareBoardComponent and click Switch', async () => {
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container: Element | undefined
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        const switchElement = container?.querySelector('.Switch')
        expect(switchElement).toBeDefined()
        userEvent.click(switchElement!)
        expect(mockedOctoClient.setSharing).toBeCalledTimes(1)
        expect(mockedOctoClient.getSharing).toBeCalledTimes(1)
        expect(container).toMatchSnapshot()
    })
    test('return shareBoardComponent and click Switch without sharing', async () => {
        mockedOctoClient.getSharing.mockResolvedValue(undefined)
        mockedUtils.createGuid.mockReturnValue('aToken')
        let container: Element | undefined
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
            mockedOctoClient.getSharing.mockResolvedValue({
                id: boardId,
                enabled: true,
                token: 'aToken',
            })
            const switchElement = container?.querySelector('.Switch')
            expect(switchElement).toBeDefined()
            userEvent.click(switchElement!)
            jest.runOnlyPendingTimers()
            result.rerender(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />))
        })

        expect(mockedOctoClient.setSharing).toBeCalledTimes(1)
        expect(mockedOctoClient.getSharing).toBeCalledTimes(2)
        expect(mockedUtils.createGuid).toBeCalledTimes(1)
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot with sharing and without workspaceId and subpath', async () => {
        w.baseURL = '/test-subpath/plugins/boards'
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        params = {
            boardId,
            viewId,
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot with sharing and subpath', async () => {
        w.baseURL = '/test-subpath/plugins/boards'
        const sharing:ISharing = {
            id: boardId,
            enabled: true,
            token: 'oneToken',
        }
        mockedOctoClient.getSharing.mockResolvedValue(sharing)
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ShareBoardComponent
                    boardId={board.id}
                    onClose={jest.fn()}
                />), {wrapper: MemoryRouter})
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
})
