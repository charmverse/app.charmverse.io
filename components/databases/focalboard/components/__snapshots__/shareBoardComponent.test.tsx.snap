// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/shareBoardComponent return shareBoardComponent and click Copy link 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/workspace/1/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/workspace/1/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copied!
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent return shareBoardComponent and click Regenerate token 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/workspace/1/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/workspace/1/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent return shareBoardComponent and click Switch 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/workspace/1/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/workspace/1/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent return shareBoardComponent and click Switch without sharing 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/workspace/1/shared/1/1?r=aToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/workspace/1/shared/1/1?r=aToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent should match snapshot 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Publish and share this board with anyone who has the link
        </div>
        <div
          class="Switch"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent should match snapshot with sharing 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/workspace/1/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/workspace/1/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent should match snapshot with sharing and subpath 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/test-subpath/plugins/boards/workspace/1/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/test-subpath/plugins/boards/workspace/1/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent should match snapshot with sharing and without workspaceId 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`src/components/shareBoardComponent should match snapshot with sharing and without workspaceId and subpath 1`] = `
<div>
  <div
    class="Modal bottom"
  >
    <div
      class="toolbar hideOnWidescreen"
    >
      <button
        aria-label="Close"
        class="Button IconButton"
        title="Close"
        type="button"
      >
        <i
          class="CompassIcon icon-close CloseIcon"
        />
      </button>
    </div>
    <div
      class="ShareBoardComponent"
    >
      <div
        class="row"
      >
        <div>
          Anyone with the link can view this board and all cards in it.
        </div>
        <div
          class="Switch on"
        >
          <div
            class="octo-switch-inner"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <a
          class="shareUrl"
          href="http://localhost/test-subpath/plugins/boards/shared/1/1?r=oneToken"
          rel="noreferrer"
          target="_blank"
        >
          http://localhost/test-subpath/plugins/boards/shared/1/1?r=oneToken
        </a>
        <button
          type="button"
        >
          <span>
            Copy link
          </span>
        </button>
      </div>
      <div
        class="row"
      >
        <button
          type="button"
        >
          <span>
            Regenerate token
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
